### PR TITLE
Fix sporadic failures in AsyncSearchAsyncTests

### LINKED ISSUE #53360: [CI] AsyncSearchActionTests fails unpredictably
We've had a couple of test failures from `AsyncSearchActionTests` today. The failures haven't reproduced locally for me.

Failure on `master`: https://gradle-enterprise.elastic.co/s/asvh7ted2nfvy/

```
org.elasticsearch.xpack.search.AsyncSearchActionTests > testInvalidId FAILED |  
-- | --
  | java.lang.AssertionError: Values should be different. Actual: 2 |  
  | at org.junit.Assert.fail(Assert.java:88) |  
  | at org.junit.Assert.failEquals(Assert.java:185) |  
  | at org.junit.Assert.assertNotEquals(Assert.java:199) |  
  | at org.junit.Assert.assertNotEquals(Assert.java:211) |  
  | at org.elasticsearch.xpack.search.AsyncSearchIntegTestCase$2.lambda$doNext$0(AsyncSearchIntegTestCase.java:214) |  
  | at org.elasticsearch.test.ESTestCase.assertBusy(ESTestCase.java:881) |  
  | at org.elasticsearch.test.ESTestCase.assertBusy(ESTestCase.java:854) |  
  | at org.elasticsearch.xpack.search.AsyncSearchIntegTestCase$2.doNext(AsyncSearchIntegTestCase.java:209) |  
  | at org.elasticsearch.xpack.search.AsyncSearchIntegTestCase$2.next(AsyncSearchIntegTestCase.java:187) |  
  | at org.elasticsearch.xpack.search.AsyncSearchIntegTestCase$2.next(AsyncSearchIntegTestCase.java:172) |  
  | at org.elasticsearch.xpack.search.AsyncSearchActionTests.testInvalidId(AsyncSearchActionTests.java:234) |  
  |   |  
  | java.lang.AssertionError: Accounting breaker not reset to 0 on node: node_sd3, are there still Lucene indices around? |  
  | Expected: <0L> |  
  | but: was <11134L> |  
  | at org.hamcrest.MatcherAssert.assertThat(MatcherAssert.java:18) |  
  | at org.junit.Assert.assertThat(Assert.java:956) |  
  | at org.elasticsearch.test.InternalTestCluster.lambda$ensureEstimatedStats$37(InternalTestCluster.java:2195) |  
  | at org.elasticsearch.test.ESTestCase.assertBusy(ESTestCase.java:881) |  
  | at org.elasticsearch.test.ESTestCase.assertBusy(ESTestCase.java:854) |  
  | at org.elasticsearch.test.InternalTestCluster.ensureEstimatedStats(InternalTestCluster.java:2193) |  
  | at org.elasticsearch.test.TestCluster.assertAfterTest(TestCluster.java:90) |  
  | at org.elasticsearch.test.InternalTestCluster.assertAfterTest(InternalTestCluster.java:2235) |  
  | at org.elasticsearch.test.ESIntegTestCase.afterInternal(ESIntegTestCase.java:550) |  
  | at org.elasticsearch.test.ESIntegTestCase.cleanUpCluster(ESIntegTestCase.java:1936) |  
  […]
```

Earlier failure on ~7.x~ master: https://gradle-enterprise.elastic.co/s/3pjzmjjxc66wi
```
org.elasticsearch.xpack.search.AsyncSearchActionTests > testCleanupOnFailure FAILED |  
-- | --
  | java.lang.AssertionError: |  
  | Expected: <6> |  
  | but: was <5> |  
  | at __randomizedtesting.SeedInfo.seed([43B69EECA92F38A8:C8A2BFFE0F2A2760]:0) |  
  | at org.hamcrest.MatcherAssert.assertThat(MatcherAssert.java:18) |  
  | at org.junit.Assert.assertThat(Assert.java:956) |  
  | at org.junit.Assert.assertThat(Assert.java:923) |  
  | at org.elasticsearch.xpack.search.AsyncSearchActionTests.testCleanupOnFailure(AsyncSearchActionTests.java:221)
```

I believe @jimczi is working on this actively, but I'm going to go ahead and mute while he's in progress.

Relates #49931

### PR DESCRIPTION
**Pull request by jimczi**
This change fixes a [race condition](https://github.com/elastic/elasticsearch/commit/8612b7541322c2635106035c31d913da645bae59) in shard group failure callbacks and ensures that we set the [correct flag on initial stored responses](https://github.com/elastic/elasticsearch/commit/da65d6f97064ac08e7bcb7a9af6bcd8a912ca7c8).

Relates #49931
Closes #53360

### CODE REVIEW

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/search/action/AsyncSearchResponse.java [hunk 1]
```diff
@@ -99,7 +99,7 @@ public void writeTo(StreamOutput out) throws IOException {
         out.writeLong(expirationTimeMillis);
     }
 
-    public AsyncSearchResponse clone(String id) {
+    public AsyncSearchResponse clone(String id, boolean isRunning) {
```
Comment (javanna) [inline x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/search/action/AsyncSearchResponse.java]: do we ever clone it providing true as isRunning argument?

File: x-pack/plugin/async-search/src/main/java/org/elasticsearch/xpack/search/TransportSubmitAsyncSearchAction.java [hunk 2]
```diff
@@ -83,7 +83,10 @@ public void onResponse(AsyncSearchResponse searchResponse) {
                                 onFatalFailure(searchTask, cause, false, submitListener);
                             } else {
                                 final String docId = searchTask.getSearchId().getDocId();
-                                store.storeInitialResponse(docId, searchTask.getOriginHeaders(), searchResponse,
+                                // creates the fallback response if the node crashes/restarts in the middle of the request
+                                // TODO: store intermediate results ?
```
Comment (javanna) [inline x-pack/plugin/async-search/src/main/java/org/elasticsearch/xpack/search/TransportSubmitAsyncSearchAction.java]: can you elaborate on this TODO? does it revolve around resiliency?

**Review by javanna — COMMENTED** [review]
I left two questions

File: x-pack/plugin/async-search/src/main/java/org/elasticsearch/xpack/search/TransportSubmitAsyncSearchAction.java (hunk 2 above)
Comment (jimczi) [reply x-pack/plugin/async-search/src/main/java/org/elasticsearch/xpack/search/TransportSubmitAsyncSearchAction.java]: It does, yes that's one of the follow up question we have in the meta issue.

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/search/action/AsyncSearchResponse.java (hunk 1 above)
Comment (jimczi) [reply x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/search/action/AsyncSearchResponse.java]: no we don't, I'll remove the isRunning and always set it to `false` on cloning if it's what you meant ?

Comment (javanna) [reply x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/search/action/AsyncSearchResponse.java]: ++

File: x-pack/plugin/async-search/src/main/java/org/elasticsearch/xpack/search/TransportSubmitAsyncSearchAction.java (hunk 2 above)
Comment (javanna) [reply x-pack/plugin/async-search/src/main/java/org/elasticsearch/xpack/search/TransportSubmitAsyncSearchAction.java]: I thought so, I wonder if we need the TODO in the code then, cause we are tracking this anyways elsewhere.

**Pull request merged**